fuse/fuse.cpp

```diff
@@ -84,25 +85,61 @@ static bool bind_mount(const std::string& source, const std::string& destination
 	return true;
 }
 
+static bool validate_mount(const std::string& op, const std::string& source)
+{
+	const std::set<std::string> mount_ops {"bind", "workspace"};
+
+	if (mount_ops.find(op) == mount_ops.end()) {
+		std::cerr << "unknown mount type: '" << op << "'" << std::endl;
+		return false;
+	}
+
+	if (op == "workspace" && source.length() != 0) {
+		std::cerr << "mount: 'workspace' can not have 'source' option" << std::endl;
+		return false;
+	}
+	return true;
+}
+
 // Do the mounts specified in the parsed input json.
 // The input/caller responsibility to ensure that the mountpoint exists,
 // that the platform supports the mount type/options, and to correctly order
 // the layered mounts.
-static bool do_mounts_from_json(const JAST& jast)
+static bool do_mounts_from_json(const JAST& jast, const std::string& fuse_mount_path)
 {
 	for (auto &x : jast.get("mounts").children) {
+		const std::string& op = x.second.get("type").value;
 		const std::string& src = x.second.get("source").value;
 		const std::string& dest = x.second.get("destination").value;
 		bool readonly = x.second.get("read-only").kind == JSON_TRUE;
 
-		if (!bind_mount(src, dest, readonly))
+		if (!validate_mount(op, src))
+			return false;
+
+		if (op == "bind" && !bind_mount(src, dest, readonly))
+			return false;
+
+		if (op == "workspace" && !bind_mount(fuse_mount_path, dest, false))
 			return false;
+
 	}
 	return true;
 }
 
 #endif
 
+static std::string get_running_dir(const JAST& jast,
+		const std::string& host_workspace_dir, const std::string& subdir)
+{
+	for (auto &x : jast.get("mounts").children) {
+		const std::string& op = x.second.get("type").value;
+		if (op == "workspace")
+			return x.second.get("destination").value + "/" + subdir;
```

Comment:
This value might be a relative path, in which case, you need to qualify it with the host_workspace_dir.